src/graphics/shadow.odin

```diff
@@ -0,0 +1,351 @@
+            la.matrix4_perspective_f32(45, f32(500) / f32(500), 0.1, 100),
+        }
+        append(&shadows.data, shadow)
+    }
+    shadows_size := len(shadows.data) * size_of(Shadow)
+    // Duplicate for frames
+    for i in 0..<frame_count { buffers_write(shadows.buffer, raw_data(shadows.data), Shadow, i * shadows_size)}
+    */
+    
+    builders.bind_descriptor_set(device, 
+                                {shadows.buffer.handle, 0, size_of(Shadow)},
+                                .UNIFORM_BUFFER_DYNAMIC,
+                                shadows.mat.descriptors[0], 0)
+    builders.bind_descriptor_set(device, 
+                                {scene.object_ubos.handle, 0, size_of(mat4)},
+                                .UNIFORM_BUFFER_DYNAMIC,
+                                shadows.mat.descriptors[3], 0)
+
+
+    shadow := Shadow {
+        la.matrix4_look_at(
+            vec3{0, 2, 0},
+            vec3{0, 0, 0},
+            vec3{0, 0, -1},
+        ),
+        //la.matrix4_translate_f32(vec3(-lights^.data[0].direction.xyz)),
+        la.matrix_ortho3d_f32(-5, 5, -5, 5, 0.1, 5),
+        //la.matrix4_perspective_f32(45, f32(500) / f32(500), 0.1, 100),
+    }
+    for i in 0..<frame_count { buffers_write(shadows.buffer, &shadow, Shadow, i)}
+}
+
+shadow_destroy :: proc(device: vk.Device, shadows: ^ShadowContext){
+    for i in 0..<shadows.frame_count {
+        vk.DestroyFramebuffer(device, shadows.framebuffers[i], nil)
+        buffers_destroy_image(device, shadows.images[i])
+    }
+    vk.DestroyRenderPass(device, shadows.render_pass, nil)
+    vk.DestroyPipeline(device, shadows.mat.effect.pipeline, nil)
+    buffers_destroy(shadows.buffer)
+}
+
+scene_render_shadows :: proc(scene: ^Scene,
+                            cmd: vk.CommandBuffer,
+                            frame_num: int,
+                            object_num: int,
+                            shadow_mat: ^Material) 
+{
+    vertex_buffer := scene.vertex_buffers[object_num]
+    index_buffer := scene.index_buffers[object_num]
+
+    model := scene.offsets(object_num, scene.time, scene.model[object_num])
+    buffers_write(scene.object_ubos,
+                  &model,
+                  mat4,
+                  frame_num * OBJECT_COUNT + object_num)	
+                  
+    // bind object uniforms
+    dynamic_offset := u32(size_of(mat4) * object_num)
+    vk.CmdBindDescriptorSets(cmd, .GRAPHICS,
+                             shadow_mat.effect.pipeline_layout, 3,
+                             1, &shadow_mat.descriptors[3],
+                             1, &dynamic_offset)
+
+    offset : vk.DeviceSize = 0
+    vk.CmdBindVertexBuffers(cmd, 0, 1, &vertex_buffer.handle, &offset)
+    vk.CmdBindIndexBuffer(cmd, index_buffer.handle, 0, .UINT16)
+    vk.CmdDrawIndexed(cmd, u32(index_buffer.size / size_of(u16)), 1, 0, 0, 0)
+}
+
+scene_draw_shadows :: proc(scene: ^Scene, cmd: vk.CommandBuffer, extent: vk.Extent2D, frame: int, obj_count: int){
```

Comment:
as it stands the difference between scene_draw_shadows and scene_render_shadows is unclear. 
I would rename this proc to scene_do_shadow_pass or something like that

Reply:
and rename scene_render_shadows proc to scene_draw_shadow